x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -406,47 +418,72 @@ public void expandIds(String idExpression,
                         if (idValue instanceof String) {
-                            foundResourceIds.add(idValue.toString());
+                            foundFromDocs.add(idValue.toString());
                         }
                     }
+                    Set<String> allFoundIds = collectIds(pageParams, foundResourceIds, foundFromDocs, totalHitCount);
                     ExpandedIdsMatcher requiredMatches = new ExpandedIdsMatcher(tokens, allowNoResources);
-                    requiredMatches.filterMatchedIds(foundResourceIds);
+                    requiredMatches.filterMatchedIds(allFoundIds);
                     if (requiredMatches.hasUnmatchedIds()) {
                         idsListener.onFailure(ExceptionsHelper.missingTrainedModel(requiredMatches.unmatchedIdsString()));
                     } else {
-                        idsListener.onResponse(Tuple.tuple(totalHitCount, foundResourceIds));
+
+                        idsListener.onResponse(Tuple.tuple(totalHitCount, allFoundIds));
                     }
                 },
                 idsListener::onFailure
             ),
             client::search);
     }
 
+    static Set<String> collectIds(PageParams pageParams, Set<String> foundFromResources, Set<String> foundFromDocs, long totalMatchedIds) {
+        TreeSet<String> allFoundIds = new TreeSet<>(foundFromDocs);
+        allFoundIds.addAll(foundFromResources);
+        int from = pageParams.getFrom();
+        int bufferedFrom = Math.min(foundFromResources.size(), from);
+
+        // If size = 10_000 but there aren't that many total IDs, reduce the size here to make following logic simpler
+        int sizeLimit = (int)Math.min(pageParams.getSize(), totalMatchedIds - from);
+
+        // Last page this means that if we "buffered" the from pagination due to resources we should clear that out
+        // We only clear from the front as that would include buffered IDs that fall on the previous page
+        if (from + sizeLimit >= totalMatchedIds) {
+            while (bufferedFrom > 0 || allFoundIds.size() > sizeLimit) {
+                allFoundIds.remove(allFoundIds.first());
+                bufferedFrom--;
+            }
+        }
+
+        // Systematically remove items while we are above the limit
+        while (allFoundIds.size() > sizeLimit) {
+            // If we are still over limit, and have buffered items, that means the first ids belong on the previous page
+            if (bufferedFrom > 0) {
+                allFoundIds.remove(allFoundIds.first());
+                bufferedFrom--;
+            } else {
+                // If we have removed all items belonging on the previous page, but are still over sized, this means we should
+                // remove items that belong on the next page.
+                allFoundIds.remove(allFoundIds.last());
+            }
+        }
+        return allFoundIds;
```

Comment:
My way of looking at it is this: we have a request for 10 ids and 2 models stored as resource. We query for 14 ids so we can figure out where the resource model ids would fit. We have a result like 

AAOOOOOOOOOOBB 

where the As and Bs are the padding.

If the model resource Ids come before the As then they are outside the ordering and all the Os are returned. Same if the model resource ids come after the Bs. 

If they are inserted middle (ids M & N) we have an ordering like

AAOOOOOMOOONOOBB

we we take the first 10 after the As. 

Same for 
    AAMOOOOOONOOOOBB
    MNAAOOOOOOOOOOBB
    AMAOOOOOONOOOOBB
    AAOOOOOOOOOOBBMN

I think we have to track the insertion position of the model resource Ids. `Collections.binarySearch()` would give the insert position. 

There is a further complication when the search does not return size hits and you have
   AAOOOOOOOOOO 
or 
    AAOOOO

Reply:
There are no `BB` at the end. moving back `from` by 2 and increasing `size` by 2 simply adds two new items to the start because it moves the page view back but it will still include the end of the initially desired page. 

If that makes sense.

Reply:
Let me take your example and run with it. Assuming size is 10.
`10 ids and 2 models`
Since we decrease `from` by two, we need to increase the `size` by two to fill out the page. If we didn't we would end up with
AAOOOOOOOO (2 As, 8 Os)
But, we increase the size to capture those two Os and our search from the docs returns

AAOOOOOOOOOO

If the resource models are inserted middle (ids M & N) we have an ordering like:

AAOOOOOMOOONOO

since our `size` is 14 in this case, we trim the first two `A`s and then the last two `O`

Returning to the user

OOOOOMOOON

Here are the results for the rest of your example insertion points
1. MOOOOOONOO
2. AAOOOOOOOO
3. AOOOOOONOO
4. OOOOOOOOOO

I think 1 and 4 are obvious OK situations. 

2 and 3 may seem weird off hand, but they are OK as well, since `M` or `N` would have been at the END of the previous page and "pushed" these two buffered IDS onto this page.

> I think we have to track the insertion position of the model resource Ids. 

I don't think this is possible since we don't know all the page boundries (without the buffers) and knowing the insertion point does not add anything.

Reply:
Thanks for the explanation. From those examples it seems once you have taken the 2 sorted sets and merged them together there are 2 conditions to consider.

1. `from == 0` in which case we remove from the back of the set until `allFoundIds.size <= pageparams.size`

In this case there are no As to remove so we trim the set to size from the back

2. `from > 0`,  remove the number of resource models from the front of the set - this is the As or another id that displaced the As. Then trim the set to size from the back. 

With those observations the code could be simplified to 
```
    static Set<String> collectIds(PageParams pageParams, Set<String> foundFromResources, Set<String> foundFromDocs) {
        // If there are no matching resource models, there was no buffering and the models from the docs
        // are paginated correctly.
        if (foundFromResources.isEmpty()) {
            return foundFromDocs;
        }

        TreeSet<String> allFoundIds = new TreeSet<>(foundFromDocs);
        allFoundIds.addAll(foundFromResources);

        if (pageParams.getFrom() > 0) {
            // not the first page so there will be extra results at the front to remove
            int numToTrimFromFront = Math.min(foundFromResources.size(), pageParams.getFrom());
            for (int i = 0; i < numToTrimFromFront; i++) {
                allFoundIds.remove(allFoundIds.first());
            }
        }

        // trim down to size removing from the rear
        while (allFoundIds.size() > pageParams.getSize()) {
            allFoundIds.remove(allFoundIds.last());
        }

        return allFoundIds;
    }
```

Reply:
@davidkyle wrote some more test cases, and this solution works great. I modified my code :D